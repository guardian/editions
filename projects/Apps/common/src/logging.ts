import { CASExpiry } from '../../../Apps/common/src/cas-expiry'
import { NetInfoStateType } from '@react-native-community/netinfo'
import { User } from '@sentry/react-native'

export enum Level {
    INFO = 'INFO',
    WARN = 'WARN',
    ERROR = 'ERROR',
    DEBUG = 'DEBUG',
}

export enum Feature {
    DOWNLOAD = 'DOWNLOAD',
    PUSH_NOTIFICATION = 'PUSH_NOTIFICATION',
    BACKGROUNG_DOWNLOAD = 'BACKGROUND_DOWNLOAD',
    CLEAR_ISSUES = 'CLEAR_ISSUES',
    SIGN_IN = 'SIGN_IN',
}

export enum ReleaseChannel {
    BETA = 'BETA',
    RELEASE = 'RELEASE',
}

export enum OS {
    IOS = 'ios',
    ANDROID = 'android',
}

export interface MallardLogFormat {
    timestamp: Date
    level: Level
    message: string
    app: string
    version: string
    buildNumber: string
    release_channel: ReleaseChannel
    selectedEdition: string
    defaultEdition: string
    os: OS
    device: string
    networkStatus: NetInfoStateType
    deviceId: string
    signedIn: boolean
    userId: User['id'] | null
    digitalSub: boolean
    casCode: CASExpiry['subscriptionCode'] | null
    iAP: boolean
    feature?: Feature
}
