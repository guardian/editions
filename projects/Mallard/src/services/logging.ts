// Logging Service that sends event logs to ELK
import NetInfo, { NetInfoStateType } from '@react-native-community/netinfo';
import type ApolloClient from 'apollo-client';
import gql from 'graphql-tag';
import { Platform } from 'react-native';
import Config from 'react-native-config';
import DeviceInfo from 'react-native-device-info';
import { getCASCode } from 'src/authentication/helpers';
import { isInBeta } from 'src/helpers/release-stream';
import type { GdprSwitchSetting } from 'src/helpers/settings';
import { defaultSettings } from 'src/helpers/settings/defaults';
import {
	iapReceiptCache,
	loggingQueueCache,
	userDataCache,
} from 'src/helpers/storage';
import {
	getDefaultEditionSlug,
	getSelectedEditionSlug,
} from 'src/hooks/use-edition-provider';
import type { MallardLogFormat } from '../../../Apps/common/src/logging';
import {
	Feature,
	Level,
	OS,
	ReleaseChannel,
} from '../../../Apps/common/src/logging';
import { AsyncQueue } from '../helpers/async-queue-cache';
import { errorService } from './errors';
import { remoteConfigService } from './remote-config';

const { LOGGING_API_KEY } = Config;
const ATTEMPTS_THEN_CLEAR = 10;
const MAX_NUM_OF_LOGS = 30;

interface LogParams {
	level: Level;
	message: string;
	optionalFields?: object;
}

type QueryData = { gdprAllowPerformance: GdprSwitchSetting };
const QUERY = gql('{ gdprAllowPerformance @client }');

const cropMessage = (message: string, maxLength: number): string => {
	return message.length > maxLength
		? `${message.slice(0, 6)}... (message cropped)`
		: message;
};

class Logging extends AsyncQueue {
	hasConsent: GdprSwitchSetting;
	numberOfAttempts: number;
	enabled: boolean;

	constructor() {
		super(loggingQueueCache);
		this.hasConsent = false;
		this.numberOfAttempts = 0;
		this.enabled = remoteConfigService.getBoolean('logging_enabled');
	}

	init(apolloClient: ApolloClient<object>) {
		apolloClient
			.watchQuery<QueryData>({ query: QUERY })
			.subscribe({
				next: (query) => {
					if (query.loading) return;
					this.hasConsent = query.data.gdprAllowPerformance;
				},
				error: (error) => {
					errorService.captureException(error);
				},
			});
	}

	async getExternalInfo() {
		const [
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		] = await Promise.all([
			NetInfo.fetch(),
			userDataCache.get(),
			getCASCode(),
			iapReceiptCache.get(),
		]);
		return {
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		};
	}

	async baseLog({
		level,
		message,
		...optionalFields
	}: LogParams): Promise<MallardLogFormat> {
		const {
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		} = await this.getExternalInfo();

		// User Data and Subscription
		const userId = userData?.userDetails.id ?? '';
		const digitalSub =
			userData?.membershipData.contentAccess.digitalPack ?? false;
		const iAP = iapReceipt ? true : false;
		const selectedEdition = await getSelectedEditionSlug();
		const defaultEdition = await getDefaultEditionSlug();

		return {
			app: DeviceInfo.getBundleId(),
			version: DeviceInfo.getVersion(),
			buildNumber: DeviceInfo.getBuildNumber(),
			os: Platform.OS === 'ios' ? OS.IOS : OS.ANDROID,
			device: DeviceInfo.getDeviceId(),
			networkStatus: networkStatus
				? networkStatus.type
				: NetInfoStateType.unknown,
			selectedEdition: selectedEdition,
			defaultEdition: defaultEdition,
			release_channel: isInBeta()
				? ReleaseChannel.BETA
				: ReleaseChannel.RELEASE,
			timestamp: new Date(),
			level,
			message,
			deviceId: DeviceInfo.getUniqueId(),
			signedIn: userData ? true : false,
			userId,
			digitalSub,
			casCode,
			iAP,
			...optionalFields,
		};
	}

	async postLogToService(log: object[]): Promise<Response | Error> {
		const response = await fetch(defaultSettings.logging, {
			method: 'POST',
			headers: {
				Accept: 'application/json',
				'Content-Type': 'application/json',
				'x-api-key': LOGGING_API_KEY,
			},
			body: JSON.stringify(log),
		});
		if (response.status !== 200) {
			throw new Error(
				`Bad response from Logging Service - status: ${response.status}`,
			);
		}
		return response;
	}

	async postLogs() {
		try {
			const { isConnected } = await NetInfo.fetch();
			// Only attempt if we are connected, otherwise wait till next time
			if (isConnected) {
				const logsToPost = await this.getQueuedItems();

				if (logsToPost.length > 1) {
					await this.postLogToService(logsToPost);
					await this.clearItems();
					this.numberOfAttempts = 0;
				}
			}
		} catch (e) {
			if (this.numberOfAttempts >= ATTEMPTS_THEN_CLEAR) {
				await this.clearItems();
				this.numberOfAttempts = 0;
			} else {
				this.numberOfAttempts++;
			}
			errorService.captureException(e);
			return e;
		}
	}

	async log({ level, message, ...optionalFields }: LogParams) {
		try {
			if (!this.hasConsent) {
				return;
			}

			// limit max length of message we post to logging service
			const croppedMessage = cropMessage(message, 300);

			const currentLog = await this.baseLog({
				level,
				message: croppedMessage,
				...optionalFields,
			});

			return this.upsertQueuedItems([currentLog], MAX_NUM_OF_LOGS);
		} catch (e) {
			errorService.captureException(e);
			return e;
		}
	}

	async basicLogInfo({
		level,
		message,
		...optionalFields
	}: LogParams): Promise<MallardLogFormat> {
		return await this.baseLog({ level, message, ...optionalFields });
	}
}

const loggingService = new Logging();

export { Level, Feature, Logging, loggingService, cropMessage };
