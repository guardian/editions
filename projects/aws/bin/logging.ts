#!/usr/bin/env node
import 'source-map-support/register'
import cdk = require('@aws-cdk/core')
import { LoggingStack } from '../lib/logging-stack'

const app = new cdk.App()
new LoggingStack(app, 'EditionsLoggingStack')
