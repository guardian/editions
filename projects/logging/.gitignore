*.d.ts
!types/*.d.ts
coverage
