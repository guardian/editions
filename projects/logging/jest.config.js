module.exports = {
    preset: 'ts-jest/presets/default',
    testMatch: ['**/__tests__/**/?(*.)+(spec|test).[jt]s?(x)'],
}
