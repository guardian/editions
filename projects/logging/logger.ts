import { configure, getLogger, addLayout } from 'log4js'

addLayout('json', () => {
    return logEvent => {
        return JSON.stringify(logEvent.data[0])
    }
})
configure({
    appenders: {
        jsonAppender: {
            type: 'stdout',
            layout: {
                type: 'json',
                separator: ',',
            },
        },
    },
    categories: {
        default: { appenders: ['jsonAppender'], level: 'info' },
    },
})
export const logger = getLogger()
