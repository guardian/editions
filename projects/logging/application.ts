import express = require('express')
import { Request, Response } from 'express'
import { logger } from './logger'
import { MallardLogFormat } from '../Apps/common/src/logging'
import sizeOf from 'object-sizeof'

const maxLogSize = parseInt(process.env.MAX_LOG_SIZE || '0')

const processLog = (rawData: MallardLogFormat[]) => {
    rawData.forEach(logData => {
        if (logData.message) {
            const elkJsonObject = {
                clientTimestamp: logData.timestamp,
                ...logData,
                // override any stage/stack/app properties included in logData
                stack: process.env.STACK,
                stage: process.env.STAGE,
                app: process.env.APP,
            }
            // let's rely on cloudwatch timestamp
            delete elkJsonObject.timestamp
            logger.info(elkJsonObject)
        } else {
            logger.info('Missing timestamp or message fields')
        }
    })
}

export const createApp = (): express.Application => {
    const app: express.Application = express()

    app.get('/healthcheck', (req: Request, res: Response) => {
        console.log('Healthcheck')
        res.send('I am the editions logger')
    })

    const logEndpointEnabled = process.env.LOG_ENDPOINT_ENABLED === 'true'
    const logEndpoint = '/log/mallard'

    if (logEndpointEnabled) {
        app.post(
            logEndpoint,
            express.json({ limit: '10mb' }),
            (req: Request, res: Response) => {
                if (req.body) {
                    const data = Array.isArray(req.body) ? req.body : [req.body]
                    const dataSize = sizeOf(data)
                    if (dataSize < maxLogSize) {
                        processLog(data)
                        res.send('Log success')
                    } else {
                        logger.error(
                            `Request body too large. Estimated size: ${dataSize}, max size: ${maxLogSize}.`,
                        )
                        res.send('Log skipped (too large)')
                    }
                } else {
                    logger.info(`Missing request body`)
                    res.status(400).send('Missing request body')
                }
            },
        )
    } else {
        app.post(logEndpoint, (req: Request, res: Response) => {
            res.send('Logging service disabled')
        })
    }

    return app
}
