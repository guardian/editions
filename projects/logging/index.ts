require('dotenv').config()

import awsServerlessExpress from 'aws-serverless-express'
import { Handler } from 'aws-lambda'
import { createApp } from './application'

const app = createApp()

export const handler: Handler = (event, context) => {
    awsServerlessExpress.proxy(
        awsServerlessExpress.createServer(app),
        event,
        context,
    )
}

if (require.main === module) {
    const port = 3132

    app.listen(port, () => {
        console.log(`Editions logging backend listening on port ${port}!`)
    })
}
