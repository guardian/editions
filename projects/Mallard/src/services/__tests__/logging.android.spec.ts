import { Level, Logging } from '../logging';
import MockDate from 'mockdate';

MockDate.set('2019-08-21');
jest.mock('src/helpers/release-stream', () => ({
	isInBeta: () => false,
}));
jest.mock('react-native/Libraries/Utilities/Platform', () => {
	const Platform = require.requireActual(
		'react-native/Libraries/Utilities/Platform',
	);
	Platform.OS = 'android';
	return Platform;
});

describe('logging service (Android and Release)', () => {
	describe('baseLog', () => {
		it('should return a log object that matches the snapshot', async () => {
			const loggingService = new Logging();
			loggingService.getExternalInfo = jest.fn().mockReturnValue({
				networkStatus: { type: 'wifi' },
				userData: {
					userDetails: { id: 'testId' },
					membershipData: {
						contentAccess: { digitalPack: true },
					},
				},
				casCode: 'QWERTYUIOP',
				iapReceipt: true,
			});
			const log = await loggingService.baseLog({
				level: Level.INFO,
				message: 'test log',
				optionalFields: { id: 'test' },
			});
			expect(log).toMatchSnapshot();
		});
	});
});
