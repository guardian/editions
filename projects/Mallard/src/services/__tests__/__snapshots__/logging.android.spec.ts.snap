// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logging service (Android and Release) baseLog should return a log object that matches the snapshot 1`] = `
Object {
  "app": "com.guardian.gce",
  "buildNumber": "666",
  "casCode": "QWERTYUIOP",
  "defaultEdition": null,
  "device": "iPad4,1",
  "deviceId": "FCDBD8EF-62FC-4ECB-B2F5-92C9E79AC7F9",
  "digitalSub": true,
  "iAP": true,
  "level": "INFO",
  "message": "test log",
  "networkStatus": "wifi",
  "optionalFields": Object {
    "id": "test",
  },
  "os": "android",
  "release_channel": "RELEASE",
  "selectedEdition": "daily-edition",
  "signedIn": true,
  "timestamp": 2019-08-21T00:00:00.000Z,
  "userId": "testId",
  "version": "3.1",
}
`;
